File: src/py/year_2023/day_09.py
from py.utils.debug import eprint


def generate_deltas(line: str):
    values = [int(x) for x in line.split()]
    sequences: list[list[int]] = [values]
    while True:
        next_lvl = [n - values[i-1] for i, n in enumerate(values[1:], start=1)]
        if not any(next_lvl):
            break
        values = next_lvl
        sequences.append(values)

    while sequences:
        yield sequences.pop()


def extrapolate_left(line: str):
    newval = 0
    for lvl, values in enumerate(generate_deltas(line)):
        newval = values[0] - newval
        if __debug__: eprint(f"Left L{lvl}: ", newval, values)
    return newval


def extrapolate_right(line: str):
    newval = 0
    for lvl, values in enumerate(generate_deltas(line)):
        newval = values[-1] + newval
        if __debug__: eprint(f"Right L{lvl}: ", values, newval)
    return newval


def solve_1(data: str):
    return sum(extrapolate_right(line) for line in data.strip().split("\n"))


def solve_2(data: str):
    return sum(extrapolate_left(line) for line in data.strip().split("\n"))


if __name__ == "__main__":
    import sys

    data = sys.stdin.read()
    print("\n[PART 1]", solve_1(data), sep="\n")
    print("\n[PART 2]", solve_2(data), sep="\n")

File: src/rs/year_2023/day_09.rs
use std::io::{self, Read};

macro_rules! iter_lines {
    ($data:ident) => {
        $data
            .trim()
            .split('\n')
            .map(|line| line.split(' ').flat_map(str::parse).collect::<Vec<i32>>())
    };
}

fn compute_deltas(nums: &Vec<i32>) -> Vec<i32> {
    assert!(nums.len() >= 2, "no delta for N<2");
    (1..nums.len()).map(|i| nums[i] - nums[i - 1]).collect()
}

enum Fold {
    Left,
    Right,
}

fn extrapolate(nums: &Vec<i32>, fold: Fold) -> i32 {
    let deltas = compute_deltas(nums);
    let (prev, sign) = match fold {
        Fold::Right => (*nums.last().unwrap_or(&0), 1),
        Fold::Left => (*nums.first().unwrap_or(&0), -1),
    };
    match deltas.iter().all(|&n| n == 0) {
        true => prev,
        false => prev + sign * extrapolate(&deltas, fold),
    }
}

fn solve_1(data: &String) -> i32 {
    iter_lines!(data).fold(0, |t, nums| t + extrapolate(&nums, Fold::Right))
}

fn solve_2(data: &String) -> i32 {
    iter_lines!(data).fold(0, |t, nums| t + extrapolate(&nums, Fold::Left))
}

fn main() {
    let mut data = String::new();
    io::stdin().read_to_string(&mut data).expect("no stdin");
    println!("{}\n{}", solve_1(&data), solve_2(&data));
}
